Pathogens/Utility/location_utils/mapbox_api_client.py

```diff
@@ -0,0 +1,128 @@
+import os
+import re
+import requests
+
+from dataclasses import dataclass
+from io import TextIOWrapper
+from Pathogens.Utility.location_utils.mapbox_response_cache import update_mapbox_response_cache, attempt_to_fetch_mapbox_response_from_cache
+from Pathogens.Utility.location_utils.country_codes import get_country_code
+from Pathogens.Utility.location_utils.lat_lng_generation_report import record_mapbox_request_in_latlng_report
+
+region_id_regex = re.compile(r'^region.([0-9]+)$')
+
+@dataclass
+class MapboxResponse:
+    center_coordinates: [float, float]
+    bounding_box: [float, float, float, float]
+    text: str
+    matching_text: str
+    region_name: str
+
+@dataclass
+class MapboxRequestParams:
+    mapbox_search_text: str
+    country_code: str
+    geocoder_data_type: str
+
+def get_mapbox_api_query_url(mapbox_request_params: MapboxRequestParams):
+    return f"https://api.mapbox.com/geocoding/v5/mapbox.places/{mapbox_request_params.mapbox_search_text}.json?" \
+          f"access_token={os.getenv('MAPBOX_API_KEY')}&types={mapbox_request_params.geocoder_data_type}&country={mapbox_request_params.country_code}"
+
+def extract_region_name_from_mapbox_api_response_context(context):
+    if(context is None):
+        return None
+    
+    region_data_in_context = next(
+        (element for element in context if element['id'] is not None and region_id_regex.match(element['id'])), None
+    )
+
+    if(region_data_in_context is None):
+        return None
+
+    return region_data_in_context.get('text', None)
+
+def parse_mapbox_response(response):
+    data = response.json()
+    if data and "features" in data and len(data['features']) > 0:
+        context = data['features'][0].get('context', None)
```

Comment:
Context is a misleading name but it comes from the mapbox API so that's why I'm keeping it consistent. This array actually contains information about the parent of the city (it's state, district, country, etc.)